ehrql/assurance.py

```diff
@@ -51,7 +51,11 @@ def validate(variable_definitions, test_data):
     input_data = []
     for patient_id, patient in test_data.items():
         for orm_class in orm_classes.values():
-            for record in patient[orm_class.__tablename__]:
+            if hasattr(orm_class, "row_id"):
+                for record in patient[orm_class.__tablename__]:
+                    input_data.append(orm_class(patient_id=patient_id, **record))
+            else:
+                record = patient[orm_class.__tablename__]
                 input_data.append(orm_class(patient_id=patient_id, **record))
```

Comment:
Appending to `input_data` in two locations seems cumbersome. We could also tidy away the loop. How about we arrange our data into a sequence, if it isn't already arranged into a sequence, and then extend `input_data`? Something like:

```suggestion
            if hasattr(orm_class, "row_id"):
                records = patient[orm_class.__tablename__]
            else:
                records = [patient[orm_class.__tablename__]]
            input_data.extend([orm_class(patient_id=patient_id, **r) for r in records])
```

Notice that the listcomp isn't necessary: we could remove the `[` and `]` and this would still work, because what remains would be a genexp. However, if the call to `orm_class` errored, then we would want to know sooner rather than later.

Reply:
I'm also not entirely sure about calling `hasattr` to tell us whether `patient[orm_class.__tablename__]` is a sequence of records or a single record. I'd like to know what a more experienced Pythonista thinks; @evansd? @inglesp?

Reply:
It's a slightly leaky abstraction, but I'd say it's fine for now.

(Ideally we'd have a function in `orm_utils.py` that told us whether a particular `orm_class` was one- or many-rows-per-patient, but there's already a `has_one_row_per_patient` function in that module (imported from `ehrql.query_model.nodes`) and so it's not clear what we'd call our new function.)

Reply:
The use of `extend` is neat.

Reply:
thanks! I added Iain's `extend` suggestion

Reply:
> Ideally we'd have a function in `orm_utils.py` that told us whether a particular orm_class was one- or many-rows-per-patient

We already have exactly this function!
https://github.com/opensafely-core/ehrql/blob/5ba76862b05ca4a1dac1a975eea8398d6012ced8/ehrql/utils/orm_utils.py#L104-L107

Reply:
Oh, we should use `table_has_one_row_per_patient` rather than `hasattr`, then!

Reply:
thanks, that's exactly what I needed (but didn't see)!

I replaced the `hasattr` with `table_has_one_row_per_patient`!